managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerFactoryImpl.java

```diff
@@ -873,23 +891,101 @@ public void getInfoFailed(ManagedLedgerException exception, Object ctx) {
         }, ctx);
     }
 
+    private CompletableFuture<Void> cleanupOffloaded(long ledgerId, UUID uuid, ManagedLedgerConfig mlConfig,
+                                     Map<String, String> offloadDriverMetadata, String cleanupReason, String name) {
+        log.info("[{}] Cleanup offload for ledgerId {} uuid {} because of the reason {}.",
+                name, ledgerId, uuid.toString(), cleanupReason);
+        Map<String, String> metadataMap = new HashMap();
+        metadataMap.putAll(offloadDriverMetadata);
+        metadataMap.put("ManagedLedgerName", name);
+
+        return Retries.run(Backoff.exponentialJittered(TimeUnit.SECONDS.toMillis(1),
+                        TimeUnit.SECONDS.toHours(1)).limit(10),
+                Retries.NonFatalPredicate,
+                () -> mlConfig.getLedgerOffloader().deleteOffloaded(ledgerId, uuid, metadataMap),
+                scheduledExecutor, name).whenComplete((ignored, exception) -> {
+            if (exception != null) {
+                log.warn("[{}] Error cleaning up offload for {}, (cleanup reason: {})",
+                        name, ledgerId, cleanupReason, exception);
+            }
+        });
+    }
+
     private void deleteManagedLedgerData(BookKeeper bkc, String managedLedgerName, ManagedLedgerInfo info,
-            DeleteLedgerCallback callback, Object ctx) {
+                                         CompletableFuture<ManagedLedgerConfig> mlConfigFuture,
+                                         DeleteLedgerCallback callback, Object ctx) {
+        final CompletableFuture<Map<Long, MLDataFormats.ManagedLedgerInfo.LedgerInfo>>
+                ledgerInfosFuture = new CompletableFuture<>();
+        store.getManagedLedgerInfo(managedLedgerName, false, null,
```

Comment:
Why do we read `ManagedLedgerInfo` from store instead of just use `info` in the parameter list?

Reply:
`info` is `org.apache.bookkeeper.mledger.ManagedLedgerInfo`, store returns `MLDataFormats.ManagedLedgerInfo.LedgerInfo` which has some additional info and used in OffloadUtils

Reply:
This still confuse me. All data in `ManagedLedgerInfo` comes directly from `MLDataFormats.ManagedLedgerInfo.LedgerInfo`. I think it's better to just sync all the info to `ManagedLedgerInfo`.

Reply:
Unfortunately ManagedLedgerInfo is part of of the public REST API

for instance here we return it (JSON encoded) to the client
https://github.com/apache/pulsar/blob/957a16b341197d6734f170a5cdcdd0500133c771/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java#L1262

We should not add everything to it.

We have this bad problem in Pulsar that we aren't always aware of what is leaking to the public APIs.

I prefer to keep the patch in this form.
And if we want to change ManagedLedgerInfo we can do it in a follow up work.

As this patch is fixing some kind of "bad problem" (because we are not deleting data that should have been deleted, that has some legal impact in some countries), this patch should be cherry-picked to active branches.
I won't add API changes in a patch that will be ported

Reply:
> Unfortunately ManagedLedgerInfo is part of of the public REST API

Sorry, I missed this. 


> I prefer to keep the patch in this form. And if we want to change ManagedLedgerInfo we can do it in a follow up work.
> 
> As this patch is fixing some kind of "bad problem" (because we are not deleting data that should have been deleted, that has some legal impact in some countries), this patch should be cherry-picked to active branches. I won't add API changes in a patch that will be ported

Make sense to me, this patch is good to go.